### PR TITLE
CBL-1232: Allow progress level to be set after replicator start

### PR DESCRIPTION
**Pull request by borrrden**
[API CHANGE] Addition of c4repl_setProgressLevel in c4Replicator.h

This change has to work its way through the various actors involved:  C4Replicator -> Replicator -> Pusher / Puller

The old method will still continue to function for the time being, but will be removed before 3.0.  It emits a warning to the logs when it is used now.

### CODE REVIEW

File: C/c4Internal.hh [hunk 1]
```diff
@@ -33,6 +33,28 @@ using namespace litecore;
 #define LOCK(MUTEX)     std::unique_lock<std::mutex> _lock(MUTEX)
 #define UNLOCK()        _lock.unlock();
 
+#if defined(__clang__)
```
Comment (bmeike) [inline C/c4Internal.hh:36]: What the heck is this?

File: C/c4_ee.gnu [hunk 2]
```diff
@@ -152,6 +152,7 @@ CBL {
 		c4repl_isDocumentPending;
 		c4repl_setOptions;
 		c4repl_setHostReachable;
+		c4repl_setProgressLevel;
 
 		c4socket_registerFactory;
 		c4socket_fromNative;
```
Comment (bmeike) [inline C/c4_ee.gnu:158]: Boy... I hope these files are all generated by a tool with a single source...

File: C/include/c4Replicator.h [hunk 3]
```diff
@@ -80,6 +80,13 @@ extern "C" {
         kC4Suspended     = 0x4          ///< If true, will not connect until unsuspended
     };
 
+    /** An enumeration of the levels of progress callbacks the replicator can provide */
+    typedef C4_ENUM(int32_t, C4ReplicatorProgressLevel) {
+        kC4ReplProgressOverall,         ///< Callback about completion and estimated total
```
Comment (bmeike) [inline C/include/c4Replicator.h:85]: Is this exactly two callbacks, one near the start that is a estimate and one after completion?

File: REST/tests/SyncListenerTest.cc [hunk 4]
```diff
@@ -320,6 +320,10 @@ TEST_CASE_METHOD(C4SyncListenerTest, "P2P Server Addresses", "[Listener]") {
     }
     
     for(const auto& interface : litecore::net::Interface::all()) {
+        if(!interface.isRoutable()) {
```
Comment (bmeike) [inline REST/tests/SyncListenerTest.cc:323]: This change is unrelated to the base topic?

File: Replicator/ReplicatorOptions.hh [hunk 5]
```diff
@@ -75,7 +75,14 @@ namespace litecore { namespace repl {
         bool skipDeleted() const  {return properties[kC4ReplicatorOptionSkipDeleted].asBool();}
         bool noIncomingConflicts() const  {return properties[kC4ReplicatorOptionNoIncomingConflicts].asBool();}
         bool noOutgoingConflicts() const  {return properties[kC4ReplicatorOptionNoIncomingConflicts].asBool();}
-        int progressLevel() const  {return (int)properties[kC4ReplicatorOptionProgressLevel].asInt();}
+        int progressLevel() const  {
+            if(properties[kC4ReplicatorOptionProgressLevel]) {
+                C4Warn("Passing in progress level via configuration is deprecated, use the setProgressLevel API");
```
Comment (bmeike) [inline Replicator/ReplicatorOptions.hh:80]: I think that comma should be either a period or a colon

File: Replicator/Worker.cc [hunk 6]
```diff
@@ -258,7 +258,14 @@ namespace litecore { namespace repl {
         }
     }
 
+    void Worker::setProgressNotificationLevel(int level) {
+        if(_progressNotificationLevel.exchange(level) == level) {
```
Comment (bmeike) [inline Replicator/Worker.cc:262]: This caches the new level and returns the old cached level?

**Review by bmeike — APPROVED** [review]
FIne... fine...

File: C/c4Internal.hh (hunk 1 above)
Comment (borrrden) [reply C/c4Internal.hh:36]: This is a way to disable compiler warnings for specific blocks of code.  In particular, Apple would complain mightily about the fact that I check for a "tautological falsehood" of the "non null" replicator being null, and then again it would (rightfully) complain about me passing in null during testing to check that it rejects it.

File: C/c4_ee.gnu (hunk 2 above)
Comment (borrrden) [reply C/c4_ee.gnu:158]: They are (See [generate_export_list.py](https://github.com/couchbase/couchbase-lite-core/blob/master/C/scripts/generate_export_list.py) )

File: C/include/c4Replicator.h (hunk 3 above)
Comment (borrrden) [reply C/include/c4Replicator.h:85]: No, it's the "running complete and total" counts callback that you get which periodically updates the states of how much progress has been made.

File: REST/tests/SyncListenerTest.cc (hunk 4 above)
Comment (borrrden) [reply REST/tests/SyncListenerTest.cc:323]: Yes, it was bugging me though because the test fails every time on my machine without it...

File: Replicator/ReplicatorOptions.hh (hunk 5 above)
Comment (borrrden) [reply Replicator/ReplicatorOptions.hh:80]: I'll keep that in mind for next time, but this message and method will be gone soon enough :)

File: Replicator/Worker.cc (hunk 6 above)
Comment (borrrden) [reply Replicator/Worker.cc:262]: Yes, that's how all atomic exchanges work as far as I know.

File: C/include/c4Replicator.h (hunk 3 above)
Comment (bmeike) [reply C/include/c4Replicator.h:85]: Ok.  We need some kind of indicator, then (... and I think this is a good place for it) of the relative costs of each of these things.

Comment (borrrden) [reply C/include/c4Replicator.h:85]: Cost in terms of what?  Time?  Complexity?

Comment (bmeike) [reply C/include/c4Replicator.h:85]: I have no idea.  That's what I'm asking for.  You need to say what makes them different and why someone would choose one instead of the other.

File: REST/tests/SyncListenerTest.cc (hunk 4 above)
Comment (borrrden) [reply REST/tests/SyncListenerTest.cc]: I took it out though, will put it back later

File: C/include/c4Replicator.h (hunk 3 above)
Comment (borrrden) [reply C/include/c4Replicator.h]: Ok I've added that

**Review by Sandychuang8 — APPROVED** [review]

**Review by jayahariv — APPROVED** [review]
